src/landscape.cpp

```diff
@@ -1347,45 +1347,42 @@ static std::tuple<bool, bool> FlowRiver(TileIndex spring, TileIndex begin, uint
 			TileIndex t = end + TileOffsByDiagDir(d);
 			if (IsValidTile(t) && !marks.contains(t) && FlowsDown(end, t)) {
 				marks.insert(t);
-				count++;
 				queue.push_back(t);
 			}
 		}
-	} while (!queue.empty());
+	}
 
 	bool main_river = false;
 	if (found) {
 		/* Flow further down hill. */
 		std::tie(found, main_river) = FlowRiver(spring, end, min_river_length);
-	} else if (count > 32) {
+	} else if (queue.size() > 32) {
 		/* Maybe we can make a lake. Find the Nth of the considered tiles. */
-		std::set<TileIndex>::const_iterator cit = marks.cbegin();
-		std::advance(cit, RandomRange(count - 1));
-		TileIndex lake_centre = *cit;
+		TileIndex lake_centre = *std::next(queue.begin(), RandomRange(static_cast<uint32_t>(queue.size())));
```

Comment:
Since `queue` is a vector, why not just use the `[]` operator?
```suggestion
		TileIndex lake_centre = queue[RandomRange(static_cast<uint32_t>(queue.size()))];
```